### ISSUE TITLE
All rules have default "error" severity

### ISSUE THREAD

**Issue by The-Penultimate-Defenestrator** [stylelint/stylelint-config-standard#23]
Feature request: 

I think it would be great if the default configuration could specify a severity for each rule, instead of just setting [`"error"` for all of them](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#severities-error--warning).

Not all rules are errors, some are minor warning (like the maximum number of blank spaces). In my opinion, the default configuration should have a severity set intelligently for each rule.

This would probably be trivial to implement by adding an extra configuration option for each rule.

**Comment by jeddy3** [comment]
@The-Penultimate-Defenestrator Thanks for following the breadcrumbs here and for posting this suggestion :)

[Elsewhere you thought](https://github.com/AtomLinter/linter-stylelint/issues/151#issue-148848529) that stylistic rules would make good candidates for warnings. Most of the rules in this sharable config are stylistic, but there are handful that are not:

- `block-no-empty`
- `color-no-invalid-hex`
- `declaration-block-no-ignored-properties`
- `declaration-block-no-shorthand-property-overrides`
- `function-linear-gradient-no-nonstandard-direction`
- `media-feature-no-missing-punctuation`
- `string-no-newline`
- `unit-no-unknown`

Is your suggestion to keep these as errors and downgrade the other rules to warnings? If not, can please you look through the configuration object yourself and list what you think is a suitable divide.

I'd then like to wait for @davidtheclark thoughts on this before going further, as it's a significant change.

I think what we need to ask ourselves is: is there value (to the consumers on this config) in establishing a warning and error classification precedence?

**Comment by The-Penultimate-Defenestrator** [comment]
Yeah, that's my suggestion. 

I think there's a definite benefit to the user. For things like the Atom package, it's decidedly more useful to have the distinction. Telling the users visually which rule violations will actually break their code versus which rule violations are purely stylistic errors is way more useful.

What reasons can you think of why *not* having default severities is good or beneficial?

**Comment by jeddy3** [comment]
> Telling the users visually which rule violations will actually break their code versus which rule violations are purely stylistic errors

That's one use of the severity levels, but *it's not the only one*. As an example, this config currently works perfectly for the following use case:

"I want my CI to fail on any stylistic errors (along with errors caught by the likes of `string-no-newline` and `color-no-invalid-hex`, and along with subtle mistakes caught by the likes of `declaration-block-no-ignored-properties` and `declaration-block-no-shorthand-property-overrides`) because I only want clean code committed to the project repo. However, I don't want the CI to fail when the linter detects someone has used an [unsupported browser feature](https://github.com/stylelint/stylelint/tree/master/src/rules/no-unsupported-browser-features)."

And as a config:

```js
// .stylelintrc
{ 
  "extends": "stylelint-config-standard"
  "rules" {
    "no-unsupported-browser-features": [ true, { "severity": "warning" }]
   }
}
```

Which, incidentally, is how I use `stylelint` :)

So, the question we're asking is... which use case is better to support out-of-the-box? 

Additionally, there are over 130 diverse rules in `stylelint`. Rules that can't be neatly categorised into *either* "break your code" *or* "stylistic" ones (e.g.  [`selector-class-pattern`](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-class-pattern/), [`unit-blacklist`](https://github.com/stylelint/stylelint/tree/master/src/rules/unit-blacklist) etc...). As such, this config was designed to provide a foundation to get started with `stylelint` and it has been designed [to be built on](https://github.com/stylelint/stylelint-config-standard#suggested-additions).

Your suggestion is very valid though and provides a great opportunity to revisit/clarify the role of this config.

I don't think there's a clear cut correct answer here though. Is there a clear benefit to severity levels in stylelint? Yes, absolutely. Is there a clear benefit to changing how rules are classified within this config? Umm, I'm not sure. I think we'll need to think carefully about that :)

**Comment by The-Penultimate-Defenestrator** [comment]
Other linters have severity levels set by default, like CSSLint. Maybe stylelint should follow their lead. Not sure, though. You certainly raise valid points as well. 

I'm currently using stylelint through the Atom package in conjunction with [remote-edit](https://atom.io/packages/remote-edit), which means I don't keep a full local copy of my project on my computer, so I don't think I can easily have a configuration file with the current setup of the atom package. That's why I'm resistant to covering this with my own config file.

**Comment by davidtheclark** [comment]
I am not in favor of differentiating between warnings and errors in this config, because I think the call is far too subjective and situation-specific. @The-Penultimate-Defenestrator you're right that some mistakes will result in actual bugs whereas others will not; but on my team, for example, I want both types to result in an error in the build. (The larger your codebase and team, I think, the more important it becomes to not allow otherwise unimportant inconsistencies to slip through the cracks.)

So I'm in favor of sticking with what we have now, which is just to leave all severities at their default level (error), and urge individuals to decide on their own which rules they want to demote to warnings.

**Comment by The-Penultimate-Defenestrator** [comment]
OK, then I'm closing this. Thanks for the help.

**Comment by jeddy3** [comment]
> I'm currently using stylelint through the Atom package in conjunction with `remote-edit`, 

@The-Penultimate-Defenestrator As an aside, this is fascinating. I had no idea there was such a thing. Are you editing files on the production server or are you accessing some kind of dev server via FTP? Do you use a CI for quality control or are you relying on editor tools? 

Either way, I hope you find a means to add a `.stylelintrc` to your setup so that you can configure stylelint to your liking.

**Comment by The-Penultimate-Defenestrator** [comment]
@jeddy3 it's not really a production server, it's a personal thing. Just works so that you can browse remote files and then pick one, and it's downloaded. Then the file is automatically uploaded with ctrl+S. It's nothing fancy, it just doesn't keep a full local copy.

**Comment by thomas-darling** [comment]
Sorry to revive this one, @controversial, but I'm with @jeddy3 on this :-)
Linting should generally produce only warnings, and errors should be reserved for issues where the code _cannot possibly be correct_. Just because someone put extra whitespace somewhere, it does not mean my code is any less correct, and it should not prevent my critical hotfix from being built and deployed. And I don't buy the argument that warnings would slip through the cracks - you'll see them every time you lint your codebase and in every CI build report. It's then a team decision how serious you are about cleaning those up, and when to do so.

I understand that some of this may be subjective, but if you won't provide sensible defaults, I would very much appreciate at least an option to change the default severity to `warning`, so I don't have to set that for (almost) every single rule - instead, we can selectively promote the rules _we_ feel represent actual errors. This is how linting works almost everywhere else :-)

**Comment by davidtheclark** [comment]
Many teams and projects deliberately fail CI on linting errors. E.g. pretty much every open source project I've poked around in that has JS linting.

> I don't buy the argument that warnings would slip through the cracks

That's kind of a silly thing to "not buy" when people only say that because they've seen warnings slip through cracks. That's exactly why I saw it, for example.

> if you won't provide sensible defaults, I would very much appreciate at least an option to change the default severity to warning

That would mean maintaining a separate stylelint-config-standard.

If @jeddy3 wants it, @jeddy3 gets it :) But I'm not keen on it myself.

**Comment by thomas-darling** [comment]
Trust me, I'm not a fan of unresolved warnings myself, and we will certainly fail on some of them :-)
But each team has their own way of dealing with those things, and I think it's only fair that the tools we use respect that. In this case, the syntax for setting the severity for each individual rule makes the configuration _really_ verbose and hard to maintain, so having the option to set the default severity would be much appreciated.

Oh, and you're right that most open source projects see all linting issues as errors, but remember that not everything is open source, with many contributors from all over the world ;-)
In our team, you can be pretty sure someone will walk over and poke you pretty hard if you don't resolve your warnings - but you would be poked _much_ harder if the site was down and your hotfix got delayed because of some insignificant whitespace ;-)

**Comment by jeddy3** [comment]
> Sorry to revive this one, @controversial, but I'm with @jeddy3 on this :-)

I'm a bit confused. Did you get those names the wrong way around? I was in favour of keeping things the way they are.

> If @jeddy3 wants it, @jeddy3 gets it :)

Ha :) I'm not in favour of maintaining another config. A suggestion of a second config might be moot though if... 

> so having the option to set the default severity would be much appreciated.

I can see how such a option would reduce the verbosity of your configuration object. There *could be* a default severity setting in the configuration object, and then the `severity` secondary option *could be* used to explicitly set something other on a per-rule basis e.g. 

```js
{
  defaultSeverity: "warning"
  rules: {
    "color-no-invalid-hex": [ true, { severity: "error" } ]
  }
}
```

However, I'm not sure how technical feasible such a option will be. Please consider creating an issue in the main stylelint repo and investigating the feasibility of doing this (via a Pull Request), as this is something that I doubt the core team will have time to spend on as it's a nice-to-have and there's a growing list of other issues.

**Comment by thomas-darling** [comment]
Yeah, sorry guys - it was very late night, and I clearly got some names mixed up there :-)

Anyways, I'm not requesting another standard config, just the option to set the default severity. I won't have time to do a pull request myself, but I think it is something you guys should seriously consider, as I'm sure im not the only one wanting it - and right now it's pretty frustrating to configure :-)

**Comment by jeddy3** [comment]
Feel free to contribute this feature when you do have time. stylelint is open source project built by contributors in their free time, and so more help is always welcomed and appreciated.

**Comment by davidtheclark** [comment]
Now that I see the `defaultSeverity` idea in a config, I think it's probably worth looking into! It might actually be fairly trivial to implement, and @thomas-darling is certainly right that it would make things easier for him and people with similar warning-vs-error preferences.
